probe/proc/cache.go

```diff
@@ -0,0 +1,123 @@
+package proc
+
+import (
+	"bytes"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+const (
+	filesCacheLen        = 1024
+	filesCacheExpiration = 60 * time.Second
+)
+
+type filesCacheEntry struct {
+	sync.RWMutex
+	file     File
+	contents *bytes.Buffer
+	creation time.Time
+}
+
+func (fce *filesCacheEntry) Pinned() bool { return fce.contents != nil }
+func (fce *filesCacheEntry) ReadInto(buf *bytes.Buffer) error {
+	fce.Lock()
+	defer fce.Unlock()
+
+	if fce.Pinned() { // use the saved contents
+		buf.Write(fce.contents.Bytes())
+		return nil
+	}
+	return fce.file.ReadInto(buf)
+}
+
+// A cache for files handles
+type filesCache struct {
+	*lru.Cache
+	proc Dir
```

Comment:
filesCache requires that all implementations of Dir are guaranteed to be safe for concurrent access by multiple goroutines. It would be good to put that requirement in the documentation for type Dir interface.